src/crimson/net/Socket.h

```diff
@@ -139,4 +144,166 @@ class Socket
+            if (seastar::engine().cpu_id() == ss.cpu) {
+              SocketRef _socket = socket.release();
+              std::ignore = seastar::with_gate(ss.shutdown_gate,
+                  [socket = std::move(_socket), peer_addr,
+                   &ss, fn_accept = std::move(fn_accept)] () mutable {
+                logger().debug("FixedCPUServerSocket({})::accept(): "
+                               "accepted peer {}", ss.addr, peer_addr);
+                return fn_accept(std::move(socket), peer_addr
+                ).handle_exception([&ss, peer_addr] (auto eptr) {
+                  logger().error("FixedCPUServerSocket({})::accept(): "
+                                 "fn_accept(s, {}) got unexpected exception {}",
+                                 ss.addr, peer_addr, eptr);
+                  ceph_abort();
+                });
+              });
+            } else {
+              logger().error("FixedCPUServerSocket({})::accept(): "
+                             "accepted on CPU#{}, should be on CPU#{}",
+                             ss.addr, seastar::engine().cpu_id(), ss.cpu);
+              ceph_abort();
+            }
+          });
+        }).handle_exception_type([&ss] (const std::system_error& e) {
+          if (e.code() == error::connection_aborted ||
+              e.code() == error::invalid_argument) {
+            logger().trace("FixedCPUServerSocket({})::accept(): stopped ({})",
+                           ss.addr, e);
+          } else {
+            throw;
+          }
+        }).handle_exception([&ss] (auto eptr) {
+          logger().error("FixedCPUServerSocket({})::accept(): "
+                         "got unexpected exception {}", ss.addr, eptr);
+          ceph_abort();
+        });
+      });
+    });
+  }
+
+  seastar::future<> shutdown() {
+    assert(seastar::engine().cpu_id() == cpu);
+    logger().debug("FixedCPUServerSocket({})::shutdown()...", addr);
+    return container().invoke_on_all([] (auto& ss) {
+      if (ss.listener) {
+        ss.listener->abort_accept();
+      }
+      return ss.shutdown_gate.close();
+    }).then([this] {
+      return reset();
+    });
+  }
+
+  seastar::future<> destroy() {
+    assert(seastar::engine().cpu_id() == cpu);
+    return shutdown().then([this] {
+      // we should only construct/stop shards on #0
+      return container().invoke_on(0, [] (auto& ss) {
+        assert(ss.handle);
+        return ss.handle->stop().finally([cleanup = std::move(ss.handle)] {});
+      });
+    });
+  }
+
+  static seastar::future<FixedCPUServerSocket*> create() {
+    auto cpu = seastar::engine().cpu_id();
+    // we should only construct/stop shards on #0
+    return seastar::smp::submit_to(0, [cpu] {
+      auto handle = std::make_unique<handle_t>();
+      return handle->start(cpu, construct_tag{}
+      ).then([handle = std::move(handle)] () mutable {
+        auto p_shard = handle.get();
+        p_shard->local().handle = std::move(handle);
+        return p_shard;
```

Comment:
> so what's the point of wrapping it as a sharded service if it's not sharded at all?

`FixedCPUServerSocket` is a sharded service because seastar framework requires to create `seastar::server_socket` and accept on all the available cores even with the "fixed" policy.

> this is confusing..
> shall we just use a plain `FixedCPUServerSocket`?

I admit this looks confusing but there are several constraints to address:
* Seastar "accept-on-all" constraint.
* The constraint that we can only manage seastar sharded data structures on core 0.

If we don't encapsulate the above constraints within `FixedCPUServerSocket` by making it a sharded service, similar implementations would be needed in `SocketMessenger`, making it unnecessarily complicated especially to messenger users. See the previous version, osd needs to create `SocketMessenger` as a sharded service, care about the order of destruction, and deallocate it on core 0 even if it is not running on 0. And `SocketMessenger` needs to create `seastar::server_socket` and accept on all cores even though it appears to operate on a fixed CPU.